gcp/api/cursor.py

```diff
@@ -0,0 +1,148 @@
+    ended: Whether this cursor is for a query that has finished returning data.
+  """
+
+  _ndb_cursor: ndb.Cursor | None = None
+  _cursor_state: _QueryCursorState = _QueryCursorState.ENDED
+  query_number: int = 1
+
+  @classmethod
+  def from_page_token(cls, page_token: str | None) -> Self:
+    """Generate a query cursor from a url safe page token."""
+    qc = cls()
+
+    if not page_token:
+      qc._cursor_state = _QueryCursorState.STARTED
+      qc._ndb_cursor = None
+      return qc
+
+    split_values = page_token.split(_METADATA_SEPARATOR, 1)
+    if len(split_values) == 2:
+      page_token = split_values[1]
+      try:
+        qc.query_number = int(split_values[0])
+      except ValueError as e:
+        raise ValueError('Invalid page token.') from e
+
+    if not page_token or page_token == _FIRST_PAGE_TOKEN:
+      qc._cursor_state = _QueryCursorState.STARTED
+      qc._ndb_cursor = None
+      return qc
+
+    qc._ndb_cursor = ndb.Cursor(urlsafe=page_token)
+    qc._cursor_state = _QueryCursorState.IN_PROGRESS
+    return qc
+
+  def update_from_iterator(self, it: ndb.QueryIterator) -> None:
+    """
+    Update the current cursor from the value of the ndb.iterator passed in.
+
+    Args:
+      it: the iterator to take the cursor position from.
+    """
+    try:
+      self._ndb_cursor = typing.cast(ndb.Cursor, it.cursor_after())
+      self._cursor_state = _QueryCursorState.IN_PROGRESS
+    except ndb_exceptions.BadArgumentError:
+      # This exception can happen when iterator has not begun iterating
+      # and it.next() is the very first element.
+      #
+      # In those cases, `cursor_after()`` would not be 'after' any element,
+      # throwing this exception.
+
+      # We represent this by setting the state to STARTED.
+      self._ndb_cursor = None
+      self._cursor_state = _QueryCursorState.STARTED
+
+  @property
+  def ndb_cursor(self) -> ndb.Cursor | None:
+    """The inner ndb cursor, could be None"""
+    if self._cursor_state == _QueryCursorState.IN_PROGRESS:
+      return self._ndb_cursor
+
+    return None
+
+  @property
+  def ended(self) -> bool:
+    """
+    Whether the cursor has finished or not.
+    """
+    return self._cursor_state == _QueryCursorState.ENDED
+
+  def url_safe_encode(self) -> str | None:
+    """
+    Create a url safe page token to pass back to the API caller.
+    """
+    cursor_part: str = ''
+    match self._cursor_state:
+      case _QueryCursorState.STARTED:
+        cursor_part = _FIRST_PAGE_TOKEN
+      case _QueryCursorState.IN_PROGRESS:
+        # Assume that IN_PROGRESS means self.cursor is always set.
+        # Loudly throw an exception if this is not the case
+        cursor_part = self._ndb_cursor.urlsafe().decode()  # type: ignore
+      case _QueryCursorState.ENDED:
+        # If ENDED, we want to return None to not include
+        # a token in the response
+        return None
+
+    if self.query_number == 0:
```

Comment:
When is this ever true, given https://github.com/google/osv.dev/pull/2514/files#r1730697012 ?

Reply:
Ah that should be 1

Reply:
Actually removed this entirely, there is no reason to not return this as part of the page token.